### PR TITLE
Feature extract model.py

### PR DESCRIPTION
**Pull request by pyapyapya**

### CODE REVIEW

**Comment by kozistr** [comment]
> @kozistr

오른쪽옆에 reviewers 에 kozistr 넣어주셈 noti 에 다른 것도 많이 와서 뭍혀서 못보는 경우도 많 ㅋㅋㅋ

File: feature_extract_model.py [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+    def __init__(self, txt_file):
+        self.txt_file = txt_file
+
+    def get_img_id(self) -> List:
+        with open(self.txt_file, 'rt', encoding='utf-8-sig') \
+                as english_file:
+            img_id: List = []
+            lines = english_file.readlines()
+            for line in lines:
+                img_id.append(line.split(' ')[:1])
```
Comment (kozistr) [inline feature_extract_model.py]: list comprehension 으로 1 줄로 써볼 수 있을 듯

File: feature_extract_model.py [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+class GetImageElements:
+    """
+    DataSet에 필요한 image의 id와 tags를 구하는 class이다.
+    """
+    def __init__(self, txt_file):
+        self.txt_file = txt_file
+
+    def get_img_id(self) -> List:
+        with open(self.txt_file, 'rt', encoding='utf-8-sig') \
```
Comment (kozistr) [inline feature_extract_model.py]: utf-8-sig 로 한 이유는?? `\ufeff` 로 시작하나 파일이??

File: feature_extract_model.py [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+        """
+        label -> multi-hot-encoding // vector size: 23545
+        :param idx: DataSet의 idx
+        :return: img의 tensor, label -> tags를 나타내는 list를 묶어놓은 list를 return
+        """
+
+        img = Image.open(os.path.join(self.img_dir, ' '.join(self.img_id[idx]) + '.jpg'))
+        img = img.convert('RGB')
+        if img.getbands()[0] == 'L':
+            img = img.convert('RGB')
```
Comment (kozistr) [inline feature_extract_model.py]: RGB 를 2번 수행하네, 한번 만 해도 될 듯?

File: feature_extract_model.py [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+        if self.transform is not None:
+            img = self.transform(img)
+
+        self.label[idx]: torch.Tensor = torch.tensor(self.label[idx])
+        return [img, self.label[idx]]
```
Comment (kozistr) [inline feature_extract_model.py]: [ ] 로 안묶어도 결과는 똑같을 듯

File: feature_extract_model.py [hunk 5]
```diff
@@ -0,0 +1,240 @@
+"""
+질문 리스트
+1. 전체적인 full-process가 맞는지?
+2. 왜 train하는 과정에서 1초에 1step이 나올정도로 느린가..?
```
Comment (kozistr) [inline feature_extract_model.py]: 1. DataLoader 에서 num_workers 가 1 이다 -> 라이젠 2600 면 8 정도 쓰면 적당할 듯. 이거 하면 몇 초 나옴?
2. 데이터가 hdd 에 있음

File: feature_extract_model.py [hunk 6]
```diff
@@ -0,0 +1,240 @@
+"""
+질문 리스트
+1. 전체적인 full-process가 맞는지?
+2. 왜 train하는 과정에서 1초에 1step이 나올정도로 느린가..?
+3. loss가 1epoch에 0.00xx... 이대로 괜찮은가.. -> 이 이유로는 23545개를 multi-hot-encoding을 진행하였기 때문에
```
Comment (kozistr) [inline feature_extract_model.py]: 아 class 가 23545 개임???? 그래서 loss 가 낮아 보일 수도 ㅇㅇ 너무 label 이 sparse 하면 training 이 잘 안되기도 함.

아마 잘못하면 모델이 label 을 다 0 으로 하면 최소 loss 에 도달하는 걸로 over-fit 할 수도 있음.

-> 클래스 수를 줄여보셈

File: feature_extract_model.py [hunk 7]
```diff
@@ -0,0 +1,240 @@
+"""
+질문 리스트
+1. 전체적인 full-process가 맞는지?
+2. 왜 train하는 과정에서 1초에 1step이 나올정도로 느린가..?
+3. loss가 1epoch에 0.00xx... 이대로 괜찮은가.. -> 이 이유로는 23545개를 multi-hot-encoding을 진행하였기 때문에
+몇 개의 1을 빼고는 다 0이라서 0으로 fitting 되는 것이 아닌가?
+4. 사실 data 자체는 25000개 중에서 몇천개 정도를 제외하면 dataset에서 단 한번의 tag를 갖는 것들임
+약간 강인하게 만들어본다고 다 tag들을 때려박아봤는데 잘못된건지?
```
Comment (kozistr) [inline feature_extract_model.py]: 음 처음부터 많이 쓰는 건 모델이 힘들어 할 수도 있음

File: feature_extract_model.py [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+
+    get_image_elements = GetImageElements(txt_file=config.setting['English_tags_path'])
+    img_id: List = get_image_elements.get_img_id()
+
+    path_train_file = config.setting['train_file_path']
+    num_train_data = cal_num_dataset(path_train_file)
+
+    nuswide_train_dataset = NusWideTrainDataSet(num_train_data, img_dir, label, img_id)
+    nuswide_train_dataloader = DataLoader(
+        dataset=nuswide_train_dataset, batch_size=32, shuffle=True, num_workers=0)
```
Comment (kozistr) [inline feature_extract_model.py]: 0 to 8

File: feature_extract_model.py [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+
+    path_test_file = config.setting['test_file_path']
+    num_test_data = cal_num_dataset(path_test_file)
+
+    nuswide_test_dataset = NusWideTestDataSet(num_test_data, img_dir, label, img_id)
+    nuswide_test_dataloader = DataLoader(
+        dataset=nuswide_test_dataset, batch_size=4, shuffle=True, num_workers=0)
```
Comment (kozistr) [inline feature_extract_model.py]: 0 to 8 pin_memory True

File: feature_extract_model.py (hunk 2 above)
Comment (pyapyapya) [reply feature_extract_model.py]: utf-8로 하니까 오류나길래 찾아보니 utf-8-sig로 해야한대서...
\ufeff 파일은 제가 몰라서 아마 아닐듯 하지만..

Comment (kozistr) [reply feature_extract_model.py]: ㅇㅎ 굳
